File: backend/src/handler/repository/get_repository_tree.rs
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
};

use gitdot_api::endpoint::get_repository_tree as api;
use gitdot_core::dto::{
    GetRepositoryTreeRequest, RepositoryAuthorizationRequest, RepositoryPermission,
};

use crate::{
    app::{AppError, AppResponse, AppState},
    dto::IntoApi,
    extract::{Principal, User},
};

#[axum::debug_handler]
pub async fn get_repository_tree(
    auth_user: Option<Principal<User>>,
    State(state): State<AppState>,
    Path((owner, repo)): Path<(String, String)>,
    Query(params): Query<api::GetRepositoryTreeRequest>,
) -> Result<AppResponse<api::GetRepositoryTreeResponse>, AppError> {
    let request = RepositoryAuthorizationRequest::new(
        auth_user.map(|u| u.id),
        &owner,
        &repo,
        RepositoryPermission::Read,
    )?;
    state
        .auth_service
        .verify_authorized_for_repository(request)
        .await?;

    let request = GetRepositoryTreeRequest::new(&repo, &owner, params.ref_name)?;
    state
        .repo_service
        .get_repository_tree(request)
        .await
        .map_err(AppError::from)
        .map(|tree| AppResponse::new(StatusCode::OK, tree.into_api()))
}

File: gitdot-web/packages/api/src/endpoint/repository/get-repository-tree.ts
import { z } from "zod";
import { RepositoryTreeResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const GetRepositoryTreeRequest = z.object({
  ref_name: z.string(),
});
export type GetRepositoryTreeRequest = z.infer<typeof GetRepositoryTreeRequest>;

export const GetRepositoryTreeResponse = RepositoryTreeResource;
export type GetRepositoryTreeResponse = z.infer<
  typeof GetRepositoryTreeResponse
>;

export const GetRepositoryTree = {
  path: "/repository/{owner}/{repo}/tree",
  method: "GET",
  request: GetRepositoryTreeRequest,
  response: GetRepositoryTreeResponse,
} as const satisfies Endpoint;
export type GetRepositoryTree = typeof GetRepositoryTree;
